### ISSUE TITLE
Multiplying single sparse chunks with sizes > 256 gives misplaced items?

### ISSUE THREAD

**Issue by ljdursi** [dask/dask#2586]
This shows up when dask.array and [sparse](https://github.com/mrocklin/sparse) interact; I don't know where the issue lives but it doesn't show up when using sparse alone.

If I use `dask.array.dot` or `tensordot` with 2d sparse chunks of size greater than 16x16 the result has values placed at incorrect coordinates.  The following demonstrates the issue; a (sparse) permutation matrix and its inverse are multiplied:

```python
import dask.array as da
import sparse
import random


def print_sparse(m, name):
    for i, j, v in zip(m.coords[0], m.coords[1], m.data):
        print name, i, j, v


def permutation_and_inverse_da(size):
    cols = range(size)
    random.shuffle(cols)
    # p . pT == I
    p = {(i, cols[i]): 1. for i in range(size)}
    pT = {(cols[i], i): 1. for i in range(size)}

    # single chunk
    a = da.from_array(sparse.COO(p), chunks=(size, size), asarray=False)
    b = da.from_array(sparse.COO(pT), chunks=(size, size), asarray=False)
    return a, b


def output_sparse_dot(a, b):
    # compute a and b and sparse-dot
    c = sparse.dot(a.compute(), b.compute())
    print "sparse.dot"
    print "coords type:",  c.coords.dtype
    print c
    print_sparse(c, 'sparse dot')


def output_dask_dot(a, b):
    c = da.dot(a, b).compute()
    print "da.dot"
    print "coords type:", c.coords.dtype
    print c
    print_sparse(c, 'da dot')


print "Should be identity matrix"
print "size = 16"
a, b = permutation_and_inverse_da(16)
output_sparse_dot(a, b)
output_dask_dot(a, b)

print "size = 17"
a, b = permutation_and_inverse_da(17)
output_sparse_dot(a, b)
output_dask_dot(a, b)
```

For size 16x16 arrays or when using sparse.dot I get the expected results
 
```
<COO: shape=(16, 16), dtype=float64, nnz=16>
da dot 0 0 1.0
da dot 1 1 1.0
da dot 2 2 1.0
da dot 3 3 1.0
da dot 4 4 1.0
da dot 5 5 1.0
da dot 6 6 1.0
da dot 7 7 1.0
da dot 8 8 1.0
da dot 9 9 1.0
da dot 10 10 1.0
da dot 11 11 1.0
da dot 12 12 1.0
da dot 13 13 1.0
da dot 14 14 1.0
da dot 15 15 1.0
```

For 17x17 with da.dot (but not sparse.dot) I get two items misplaced:
```
<COO: shape=(17, 17), dtype=float64, nnz=17>
da dot 0 0 1.0
da dot 0 14 1.0   # <--
da dot 1 1 1.0
da dot 1 15 1.0    # <-- 
da dot 2 2 1.0
da dot 3 3 1.0
da dot 4 4 1.0
da dot 5 5 1.0
da dot 6 6 1.0
da dot 7 7 1.0
da dot 8 8 1.0
da dot 9 9 1.0
da dot 10 10 1.0
da dot 11 11 1.0 
da dot 12 12 1.0
da dot 13 13 1.0
da dot 14 14 1.0
```

I can't help but notice that the default coo.coords dtype is `uint8` and the item at (0,14) - linear index 14 - should presumably occur at (15,15) - linear index 15*17+15 = 270, e.g. is off by 256, as with the item at (1,15) with should be at (16, 16).  Similarly, trying a 23x23 matrix (so size > 512) leads to minor diagonals, items off by 256 and items off by 512.  Performing the same calculations with a.coords and b.coords as uint32 or uint64 does not resolve the problem.

```bash
$ python
Python 2.7.13 (default, Dec 18 2016, 07:03:39)
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import dask
>>> import sparse
>>> print dask.__version__
0.15.1
>>> print sparse.__version__
0.1.0
```

**Comment by ljdursi** [comment]
Sorry, this must have already been addressed, I see that with the version of sparse in master this is resolved.

**Comment by mrocklin** [comment]
I've just released master under a new version 0.1.1.   Thank you for
reporting this.
